File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        stack = []
        pairs = {')': '(', ']': '[', '}': '{'}

        for ch in s:
            if ch in ('(', '[', '{'):
                stack.append(ch)
            else:
                if not stack or stack[-1] != pairs.get(ch):
                    return False
                stack.pop()
        return not stack

            
        

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for(int i=0; i<s.size(); i++){
            //opening case
            if(s[i] == '[' || s[i] == '{' || s[i] == '('){
                st.push(s[i]);
            }
            //closing
            else {
                if(st.size() == 0){
                    return false; //opening but no closing
                }
                if(st.top() == '(' && s[i] == ')'
                || st.top() == '[' && s[i] == ']'
                || st.top() == '{' && s[i] == '}'){
                st.pop();
            } else return false;


            }
            
        }

        return st.size() == 0;
        
    }
};